uboot-mediatek: add support for Xiaomi Redmi Router AX6000

HZFrodo opened the pull request:
uboot-mediatek: add support for Xiaomi Redmi Router AX6000

U-Boot flash instructions:

0. OpenWrt U-Boot does not support stock layout, it comes with recovery
boot support, automatic tftp recovery and never blocks UART.

A new flash layout is introduced, we call it OpenWrt U-Boot layout,
stock flash layout and the old OpenWrt layout are not supported.

During the whole flash procedure, please do not reboot or power off
unless requested explicitly, or you will break your device.

1. Your device should already running OpenWrt.
If not, follow the instructions to flash OpenWrt:
https://github.com/openwrt/openwrt/pull/11115

2. Backup BL2 Nvram Bdata Factory and FIP in case you break something or
in case you want to go back to stock firmware one day.

cat /dev/mtdblock0 > /tmp/BL2.bin
cat /dev/mtdblock1 > /tmp/Nvram.bin
cat /dev/mtdblock2 > /tmp/Bdata.bin
cat /dev/mtdblock3 > /tmp/Factory.bin
cat /dev/mtdblock4 > /tmp/FIP.bin

And save all whose bin files to somewhere safe.

Then backup your configurations, since ubiformat for entire mtd device is
required to create new ubootenv volume for OpenWrt U-Boot.

3. Run the following cmd to boot into an initramfs with the new OpenWrt
U-Boot layout that expand ubi partion to the end of flash:

ubiformat /dev/mtd7 -y -f /tmp/ax6000-ubootmod-initramfs-factory.ubi

4. After boot into initramfs, check mtd partion info.
The ubi partion should be mtd5

root@OpenWrt:~# cat /proc/mtd
dev:    size   erasesize  name
mtd0: 00100000 00020000 BL2
mtd1: 00040000 00020000 Nvram
mtd2: 00040000 00020000 Bdata
mtd3: 00200000 00020000 Factory
mtd4: 00200000 00020000 FIP
mtd5: 07a80000 00020000 ubi

5. Load kmod-mtd-rw to temporarily make the bootloader partions writable.
The kmod-mtd-rw is from the feeds, it is not packed in initramfs-kernel
by default.

To install kmod-mtd-rw via opkg

opkg update && opkg install kmod-mtd-rw

Or, download kmod-mtd-rw.ipk from OpenWrt server and install it manually
e.g:
https://downloads.openwrt.org/snapshots/targets/mediatek/filogic/kmods/
Select your OpenWrt release version and kernel version accordingly.

Load kmod-mtd-rw:
insmod /lib/modules/$(uname -r)/mtd-rw.ko i_want_a_brick=1

6. Run the following cmd to clean all pending crash dumps in pstore,
or OpenWrt U-Boot may boot into NAND recovery or tftp recovery.

rm -f /sys/fs/pstore/*

7. Format ubi and create new ubootenv volume:

ubidetach -p /dev/mtd5; ubiformat /dev/mtd5 -y; ubiattach -p /dev/mtd5
ubimkvol /dev/ubi0 -n 0 -N ubootenv -s 128KiB
ubimkvol /dev/ubi0 -n 1 -N ubootenv2 -s 128KiB

8. This is optional. Skip this if you do not want to have NAND recovery
boot feature offered by OpenWrt U-Boot. Don't worry, you always have
automatic tftp recovery feature enabled.

ubimkvol /dev/ubi0 -n 2 -N recovery -s 10MiB
ubiupdatevol /dev/ubi0_2 /tmp/ax6000-ubootmod-initramfs-recovery.itb

9. Now, flash new U-Boot. Bye-bye ugly stock U-Boot.

mtd write /tmp/ax6000-ubootmod-preloader.bin BL2
mtd write /tmp/ax6000-ubootmod-bl31-uboot.fip FIP

10. Flash the squashfs-sysupgrade.bin as usual

sysupgrade -n /tmp/ax6000-ubootmod-squashfs-sysupgrade.itb

Enjoy!

dangowrt commented on package/boot/uboot-mediatek/Makefile:
```diff
@@ -236,6 +236,18 @@ define U-Boot/mt7986_bananapi_bpi-r3-nor
   FIP_COMPRESS:=1
 endef
 
+define U-Boot/mt7986_xiaomi_redmi-router-ax6000-snand
```
The AX6000 *always* boots from SNAND. Hence there is no need to carry the word `snand` in the name.

HZFrodo replied:
Done. Thanks

dangowrt reviewed (changes requested):
Looks good apart from `-snand` in the name of the U-Boot build. Remove that and we are good to go.

dangowrt commented:
Please review our submission guideline https://openwrt.org/submitting-patches#submission_guidelines

The commit message to have line-breaks after max. 75 characters.

HZFrodo commented on target/linux/mediatek/dts/mt7986a-xiaomi-redmi-router-ax6000.dtsi:
```diff
@@ -149,7 +149,6 @@
 			partition@0 {
 				label = "BL2";
 				reg = <0x0 0x100000>;
-				read-only;
```
Done. Thanks

graysky2 commented:
@HZFrodo - Following step 1 will give a device using the stock layout by definition, so step 3 is confusing to me:
> 3. Boot into initramfs with OpenWrt layout, not the stock layout:

dangowrt commented on package/boot/uboot-mediatek/patches/439-add-xiaomi_redmi-ax6000.patch:
```diff
@@ -0,0 +1,388 @@
++bootfile=openwrt-mediatek-filogic-xiaomi_redmi-router-ax6000-initramfs-kernel.bin
++bootfile_bl2=openwrt-mediatek-filogic-xiaomi_redmi-router-ax6000-snand-preloader.bin
++bootfile_fip=openwrt-mediatek-filogic-xiaomi_redmi-router-ax6000-snand-bl31-uboot.fip
++bootfile_upg=openwrt-mediatek-filogic-xiaomi_redmi-router-ax6000-squashfs-sysupgrade.bin
```
This doesn't seem to be used anywhere, so it can be removed. As we keep using `sysupgrade.tar` format it is difficult to support flashing in U-Boot...

HZFrodo replied:
Done. Thanks

HZFrodo commented:
> Might be more clear with slightly different wording such as:
> 
> _3. Run the following command to boot into an initramfs with OpenWrt layout:_

Yes, you are right, thanks for your correction.

HZFrodo commented:
> @HZFrodo - Following step 1 will give a device using the stock layout by definition, so step 3 is confusing to me:
> 
> > 3. Boot into initramfs with OpenWrt layout, not the stock layout:
> 
> Might be more clear with slightly different wording such as:
> 
> _3. Run the following command to boot into an initramfs with OpenWrt layout:_
> 
> For step 6, could you include the needed commands rather than linking to the external webiste?
> 
> > 6. Load kmod-mtd-rw to temporarily make the bootloader partions writable
> >    see: https://github.com/jclehner/mtd-rw
> 
> Also, a dumb question perhaps, but is the needed module included in the image we used in step 3?
> 
> > sysupgrade -F /tmp/ax6000-initramfs-kernel.bin

OK, I will update a more detailed instructions, thanks.

graysky2 commented:
@HZFrodo - thanks, I think the more verbose commit message will help avoid confusion :+1:

981213 commented:
You need to consider the bmt/nmbm nonsense xD. OpenWrt u-boot doesn't support it.
nmbm should be completely disabled in the variant using OpenWrt u-boot, and you can expand ubi all the way to the end of the flash. Also, I think it's a good idea to reclaim the crash and crash_log spaces and merge them into ubi.
Just add a ubootmod suffix to the redmi-router-ax6000 variant and modify the dts. I don't want to keep the version using xiaomi u-boot and modified ubi layout. Users who want those extra space should go with a u-boot mod.

Also, because of the nmbm bad block remapping, you need to first check if there are skipped bad blocks between 0x0~0x580000. (There is a debug interface in current bmt code to dump these info but I don't have a device to test atm.)
I think the following steps should be added in the flash instruction:
1. (your original instruction)
2. (your original instruction)
3. Backup the Factory partition in stock firmware. This is WiFi calibration data and nobody can really save you if you lose it. Also backup BL2, FIP and bdata if you wish to go back to the xiaomi firmware one day.
4. Check the bad block situation: If there are skipped bad blocks in 0x0\~0x180000(i.e. before Factory), you'll need to reflash Factory partition content backed up in step 1 during u-boot replacement. If there is any bad block within 0x180000\~0x380000 (Factory), unfortunately your device isn't suitable for a u-boot mod. Stay with the stock layout.
5. (continue to the initramfs flashing step)

981213 commented:
Maybe also consider using the external data fip by @dangowrt instead of sysupgrade-tar. We are replacing all the software components made by Xiaomi so I think it's better to use modern stuff instead.

HZFrodo commented:
> Maybe also consider using the external data fit by @dangowrt instead of sysupgrade-tar. We are replacing all the software components made by Xiaomi so I think it's better to use modern stuff instead.

Yes, I agree.
I was go to use the external data fit instead of sysupgrade-tar, but I thought it will be too confusing for one device to have 3 different flash layouts and 3 different sysupgrade images. So I chosed to stay the OpenWrt layout.
If OpenWrt community will accept a new layout for the ubootmod, I am happy with it.

981213 commented:
> but I thought it will be too confusing for one device to have 3 different flash layouts and 3 different sysupgrade images.

I'm planning to drop the one using Xiaomi u-boot with modified bootcmd if this PR is merged, so there will only be two left.
Xiaomi u-boot doesn't take any UART input with certain env settings. If the u-boot is in a state without UART access and user accidentally break the firmware during flashing, they have no way to recover their router without some soldering work and a spi nand programmer. Anyone who is willing to take this risk for bigger space should go with a u-boot mod, because it isn't any more risky (Worst case scenario you flashed the wrong file and you can recover it by taking the flash off the board.) and the risk in u-boot mod is only during u-boot replacement, while the Xiaomi u-boot can relock its UART on env corruption.

HZFrodo commented:
> because of the nmbm bad block remapping, you need to first check if there are skipped bad blocks between 0x0~0x380000. (There is a debug interface in current bmt code to dump these info but I don't have a device to test atm.)

@981213 After some code digging, I found that there is a hard dependency:
1. OpenWrt kernel will not create the NMBM instance by default, since there is no valid signature
https://github.com/openwrt/openwrt/blob/master/target/linux/generic/files/drivers/mtd/nand/mtk_bmt_nmbm.c#L2179 /sys/kernel/debug/mtk-bmt/ dir will not be created.
3. Add a dts property to force create NMBM instance
https://github.com/openwrt/openwrt/blob/master/target/linux/generic/files/drivers/mtd/nand/mtk_bmt_nmbm.c#L2264
/sys/kernel/debug/mtk-bmt/ comes out, but no useful info at all. Because it is just initialized.
How can I know the original bad block info?
Maybe from the stock kernel?
My device cannot back to stock firmware anymore, I have no backup.
Could someone try this out?

HZFrodo commented:
External data fit work finished and tested.

The flash instructions updated, but the bad block check stuff is still unclear.
Could someone make a new PR to do some dts mod or driver mod and describe the bad block check procedure clearly?
Thanks.

HZFrodo commented:
Tried several ways to confirm the bad block check procedure, I still can not figure out how to check it right.

@dangowrt is also working on U-Boot support for TP-Link TL-XDR608x that has the same chip as Redmi AX6000
https://git.openwrt.org/?p=openwrt/staging/dangole.git;a=summary

@dangowrt Do you have any idea?

981213 commented on package/boot/uboot-mediatek/patches/439-add-xiaomi_redmi-ax6000.patch:
```diff
@@ -0,0 +1,391 @@
++
++	spi_nand@1 {
++		compatible = "spi-nand";
++		reg = <1>;
```
Why is SPI NAND on CS1? Is this a typo?

HZFrodo replied:
Fixed. Thanks.

981213 reviewed (commented):
Hi! some comments on the code below:

dangowrt commented on target/linux/mediatek/image/filogic.mk:
```diff
@@ -157,3 +157,31 @@ endif
+  UBINIZE_OPTS := -E 5
+  BLOCKSIZE := 128k
+  PAGESIZE := 2048
+  KERNEL_IN_UBI := 1
```
```suggestion
  KERNEL_IN_UBI := 1
  UBOOTENV_IN_UBI := 1
```
This will make sure ubinized image also contains `ubootenv` and `ubootenv2` volumes.

HZFrodo replied:
Done. Thanks.

dangowrt commented on target/linux/mediatek/image/filogic.mk:
```diff
@@ -157,3 +157,31 @@ endif
+  BLOCKSIZE := 128k
+  PAGESIZE := 2048
+  KERNEL_IN_UBI := 1
+  KERNEL := kernel-bin | gzip
```
I did this also for other boards because gzip decompression is significantly faster. Hence, for production image I wanted to minimize boot time, while for recovery image I wanted to minimize required space (and hence you lzma instead of gzip).

981213 replied:
Oh, that makes sense. There are plenty of space on NAND anyway.

HZFrodo committed:
mediatek: add new layout for Xiaomi Redmi Router AX6000 for OpenWrt U-Boot

This new layout is only bootable with OpenWrt U-Boot. It reuses the
two crash partions and expands the ubi partion to the end of whole flash.

Do not use this layout with stock U-Boot!

Signed-off-by: Furong Xu <xfr@outlook.com>

HZFrodo committed:
uboot-mediatek: add support for Xiaomi Redmi Router AX6000

U-Boot flash instructions:

0. OpenWrt U-Boot does not support stock layout, it comes with recovery
boot support, automatic tftp recovery and never blocks UART.

A new flash layout is introduced, we call it OpenWrt U-Boot layout,
stock flash layout and the old OpenWrt layout are not supported.

During the whole flash procedure, please do not reboot or power off
unless requested explicitly, or you will break your device.

1. Your device should already running OpenWrt.
If not, follow the instructions to flash OpenWrt:
https://github.com/openwrt/openwrt/pull/11115

2. Backup BL2 Nvram Bdata Factory and FIP in case you break something or
in case you want to go back to stock firmware one day.

cat /dev/mtdblock0 > /tmp/BL2.bin
cat /dev/mtdblock1 > /tmp/Nvram.bin
cat /dev/mtdblock2 > /tmp/Bdata.bin
cat /dev/mtdblock3 > /tmp/Factory.bin
cat /dev/mtdblock4 > /tmp/FIP.bin

And save all whose bin files to somewhere safe.

Then backup your configurations, since ubiformat for entire mtd device is
required to create new ubootenv volume for OpenWrt U-Boot.

3. Run the following cmd to boot into an initramfs with the new OpenWrt
U-Boot layout that expand ubi partion to the end of flash:

ubiformat /dev/mtd7 -y -f /tmp/ax6000-ubootmod-initramfs-factory.ubi

4. After boot into initramfs, check mtd partion info.
The ubi partion should be mtd5

root@OpenWrt:~# cat /proc/mtd
dev:    size   erasesize  name
mtd0: 00100000 00020000 BL2
mtd1: 00040000 00020000 Nvram
mtd2: 00040000 00020000 Bdata
mtd3: 00200000 00020000 Factory
mtd4: 00200000 00020000 FIP
mtd5: 07a80000 00020000 ubi

5. Load kmod-mtd-rw to temporarily make the bootloader partions writable.
The kmod-mtd-rw is from the feeds, it is not packed in initramfs-factory
by default.

To install kmod-mtd-rw via opkg:

opkg update && opkg install kmod-mtd-rw

Or, download kmod-mtd-rw.ipk from OpenWrt server and install it manually
e.g:
https://downloads.openwrt.org/snapshots/targets/mediatek/filogic/kmods/
Select your OpenWrt release version and kernel version accordingly.

Load kmod-mtd-rw:
insmod /lib/modules/$(uname -r)/mtd-rw.ko i_want_a_brick=1

6. Run the following cmd to clean all pending crash dumps in pstore,
or OpenWrt U-Boot may boot into NAND recovery or tftp recovery.

rm -f /sys/fs/pstore/*

7. Format ubi and create new ubootenv volume:

ubidetach -p /dev/mtd5; ubiformat /dev/mtd5 -y; ubiattach -p /dev/mtd5
ubimkvol /dev/ubi0 -n 0 -N ubootenv -s 128KiB
ubimkvol /dev/ubi0 -n 1 -N ubootenv2 -s 128KiB

8. This is optional. Skip this if you do not want to have NAND recovery
boot feature offered by OpenWrt U-Boot. Don't worry, you always have
automatic tftp recovery feature enabled.

ubimkvol /dev/ubi0 -n 2 -N recovery -s 10MiB
ubiupdatevol /dev/ubi0_2 /tmp/ax6000-ubootmod-initramfs-recovery.itb

9. Now, flash new U-Boot. Bye-bye ugly stock U-Boot.

mtd write /tmp/ax6000-ubootmod-preloader.bin BL2
mtd write /tmp/ax6000-ubootmod-bl31-uboot.fip FIP

10. Flash the squashfs-sysupgrade.bin as usual:

sysupgrade -n /tmp/ax6000-ubootmod-squashfs-sysupgrade.itb

Enjoy!

Signed-off-by: Furong Xu <xfr@outlook.com>

HZFrodo commented on target/linux/mediatek/dts/mt7986a-xiaomi-redmi-router-ax6000.dtsi:
```diff
@@ -181,18 +177,6 @@
-				label = "crash_log";
-				reg = <0x5c0000 0x40000>;
-				read-only;
-			};
```
Done. Thanks.

HZFrodo commented on target/linux/mediatek/image/filogic.mk:
```diff
@@ -157,3 +157,31 @@ endif
+  ARTIFACT/preloader.bin := bl2 spim-nand-ddr4
+  ARTIFACT/bl31-uboot.fip := bl31-uboot xiaomi_redmi-router-ax6000
+ifneq ($(CONFIG_TARGET_ROOTFS_INITRAMFS),)
+  ARTIFACTS += initramfs-ubi-exp.ubi
```
Done.

dangowrt commented:
Pushed to master https://github.com/openwrt/openwrt/commit/8bac5dc18a139e13338956e83d8dc4948824a82f https://github.com/openwrt/openwrt/commit/1613e3340b829ea9aa6da954bf0ff98214b71751

noskoski commented:
Hi did you have a full dump of the flash ? ( With new U-boot and openwrt ) Some way I corrupted everything, I will need rewrite it all
